File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        
	int ans =0, lo =0, hi = height.size()-1, loMax=height[0], hiMax = height[hi];

	while(lo <hi){
		if(height[lo] <height[hi]){
			if(loMax <= height[lo]){
				loMax = height[lo];
			}
			else{
			  ans += loMax - height[lo];
		  }
      lo++;
    }
		else{
			if(hiMax <= height[hi]){
				hiMax = height[hi];
			}
			else{
			ans += hiMax - height[hi];
		  }
      hi--;
	  }
  }
  return ans;
  }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int left = 0, right = height.length-1;
        int count =0;
        int leftMax=height[0];
        int rightMax = height[right];
        
        while(left<right){
        // System.out.println("left " + height[left] + " leftMax " + leftMax + " right " + height[right] + " rightMax " + rightMax + " count " + count);
                leftMax = Math.max(leftMax, height[left]);
                rightMax = Math.max(rightMax, height[right]);
                int currentLevel = Math.min(leftMax, rightMax);


            if(height[left]<height[right]){
                
                
                if(height[left]< currentLevel)
                {
                    count += (currentLevel-height[left]);
                }
                left++;
            }
            else{                
                if(height[right] < currentLevel)
                count += (currentLevel-height[right]);
                
                right--;
            }
        }
        return count;
    }
}